File: mobile/MedAI/composeApp/src/commonMain/kotlin/org/example/project/data/remote/dto/doctor/DoctorResponseDto.kt
package org.example.project.data.remote.dto.doctor

import kotlinx.serialization.Serializable

@Serializable
data class DoctorResponseDto(
    val userId: Int,
    val name: String? = null,
    val specialities: List<DoctorSpecialityResponseDto> = emptyList()
)

@Serializable
data class DoctorSpecialityResponseDto(
    val id: Int,
    val isPrimary: Boolean,
    val yearsOfExperience: Int,
    val speciality: SpecialityItemResponseDto? = null
)

@Serializable
data class SpecialityItemResponseDto(
    val id: Int,
    val name: String
)

File: backend/src/doctors/dtos/doctor-response.dto.ts
import { ApiProperty } from '@nestjs/swagger';
import { Doctor } from '../entities/doctor.entity';
import { DoctorSpecialityResponseDto } from './doctor-speciality-response.dto';

export class DoctorResponseDto {
  @ApiProperty()
  userId: number;

  @ApiProperty({ required: false })
  name?: string;

  @ApiProperty({ type: [DoctorSpecialityResponseDto], required: false })
  specialities?: DoctorSpecialityResponseDto[];

  @ApiProperty({ required: false })
  about?: string;

  @ApiProperty()
  rating: number = 0;

  @ApiProperty({ required: false })
  bestReview?: string = 'Not rated yet';

  constructor(partial: Partial<Doctor>) {
    this.userId = partial.userId as number;
    this.name = partial.user?.name;
    this.about = partial.about;
    if (partial.specialities) {
      this.specialities = partial.specialities.map(
        (s) => new DoctorSpecialityResponseDto(s),
      );
    }
    if (partial.appointments) {
      const rated = partial.appointments.filter(
        (a): a is typeof a & { rating: number } => a.rating !== null,
      );
      if (rated.length > 0) {
        this.rating =
          rated.reduce((sum, a) => sum + a.rating, 0) / rated.length;
        const best = rated.reduce((best, a) =>
          a.rating > best.rating ? a : best,
        );
        if (best.review) {
          this.bestReview = best.review;
        }
      }
    }
  }
}
